app/src/main/java/com/nutomic/syncthingandroid/model/Completion.java

```diff
@@ -0,0 +1,175 @@
+        Boolean deviceFound;
+        for (String deviceId : deviceFolderMap.keySet()) {
+            deviceFound = false;
+            for (Device device : newDeviceSet) {
+                if (device.deviceID.equals(deviceId)) {
+                    deviceFound = true;
+                    break;
+                }
+            }
+            if (!deviceFound) {
+                removedDevices.add(deviceId);
+            }
+        }
+        for (String deviceId : removedDevices) {
+            Log.v(TAG, "updateFromConfig: Remove device '" + deviceId + "' from cache model");
+            deviceFolderMap.remove(deviceId);
+        }
+
+        // Handle devices that were added to the config.
+        for (Device device : newDeviceSet) {
+            if (!deviceFolderMap.containsKey(device.deviceID)) {
+                Log.v(TAG, "updateFromConfig: Add device '" + device.deviceID + "' to cache model");
+                deviceFolderMap.put(device.deviceID, defaultDevice);
+            }
+        }
+
+        // Handle folders that were removed from the config.
+        List<String> removedFolders = new ArrayList<>();;
+        Boolean folderFound;
+        for (String deviceId : deviceFolderMap.keySet()) {
+            for (String folderId : deviceFolderMap.get(deviceId).keySet()) {
+                folderFound = false;
+                for (Folder folder : newFolderSet) {
+                    if (folder.id.equals(folderId)) {
+                        folderFound = true;
+                        break;
+                    }
+                }
+                if (!folderFound) {
+                    removedFolders.add(folderId);
+                }
+            }
+        }
+        for (String folderId : removedFolders) {
+            Log.v(TAG, "updateFromConfig: Remove folder '" + folderId + "' from cache model");
+            removeFolder(folderId);
+        }
+
+        // Handle folders that were added to the config.
+        for (Folder folder : newFolderSet) {
+            for (Device device : newDeviceSet) {
+                if (folder.getDevice(device.deviceID) != null) {
+                    // folder is shared with device.
+                    if (!deviceFolderMap.get(device.deviceID).containsKey(folder.id)) {
+                        Log.v(TAG, "updateFromConfig: Add folder '" + folder.id +
+                                    "' shared with device '" + device.deviceID + "' to cache model.");
+                        deviceFolderMap.get(device.deviceID).put(folder.id, defaultFolder);
+                    }
+                }
+            }
+        }
+    }
+
+    /**
+     * Calculates remote device sync completion percentage across all folders
+     * shared with the device.
+     */
+    public int getDeviceCompletion(String deviceId) {
+        int folderCount = 0;
+        double sumCompletion = 0;
+        for (String folderId : deviceFolderMap.getOrDefault(deviceId, defaultDevice).keySet()) {

```

Comment:
You are not using this correctly.

I suggest you go back to:
```
Map<String, CompletionInfo> completion = deviceFolderMap.get(deviceId);
if (completion == null) {
    return 0
}
// Don't know exact Java API but there must be a way to iterate key value pairs.
for (Pair<String, CompletionInfo> pair : completion.iterator()) {
  calculate compeltion
}
```

Reply:
Well, you wanted it short and simple by getOrDefault, now it is: Let me explain it please. When a default device is returned because of a deviceId not in the model cache, it has no keyset of folders in it, so the for won't do anything. This will lead to assume the default return value of 100% - as "nutomic" had it in the connections UI code before. It was tested working. If not okay, I would go back to the check with containsKey and "auto-learning" mode to add missing deviceId to the HashMap.

Reply:
I envisioned a different DefaultHashMap, which added the entry to the map and then returned if it was not present, avoiding these puts etc, but to be honest, it's probably better we don't rely on that magical insert if not exists behavior.

In this case it's easier just to do a get and do a null check rather than having a magic class that has one function that returns an empty object that has to be defined on the module scope.

Also the getOrDefault+get seems wrong, you could do a get once, null check, and then get the folder.